### ISSUE TITLE
bug: ion-datetime custom button with role="cancel" still updates value

### ISSUE THREAD

**Issue by simonhaenisch** [ionic-team/ionic-framework#20048]
# Bug Report

**Ionic version:**
[x] **4.x**

**Current behavior:**

I use the `pickerOptions` property of the `ion-datetime` to add buttons with custom handlers. I add a "Cancel" button with `role: 'cancel'`. When I click that cancel button, the value of the `ion-datetime` changes anyway.

**Expected behavior:**

A picker button with role cancel should not change the `ion-datetime`'s value.

**Steps to reproduce:**

```html
<ion-datetime display-format="YYYY" value="2019" />
```

```js
document.querySelector('ion-datetime').pickerOptions = {
  buttons: [
    { text: 'Cancel', role: 'cancel' },
    { text: 'Done', handler: columns => console.log(columns.year.value) },
  ],
};
```

<img src="https://user-images.githubusercontent.com/3410759/70451460-b2358780-1aa5-11ea-89ab-ccc15e7dbbcc.gif" width="366">

**Related code:**

https://codepen.io/simonhaenisch/pen/YzPwYZa?editors=1010

**Other information:**

I think #19441 is about the same issue but not sure.

This is the code that looks like it should be working:

https://github.com/ionic-team/ionic/blob/6e0b9c45489889266620ee2ca38c33fdf8ce3f3b/core/src/components/picker/picker.tsx#L166-L170

**Ionic info:**

```
@ionic/core: 4.11.5
```

**Comment by simonhaenisch** [comment]
I just noticed that because doing a backdrop-dismiss clicks the button with `'cancel'`role programmatically, the backdrop dismiss also has the same buggy behavior that it updates the value even though it should not.

**Comment by brandyscarney** [comment]
Thank you for the issue! I confirmed this is also an issue in Ionic 5's latest beta.

**Comment by brandyscarney** [comment]
I took a quick look into this, and found that the cause is that every time the picker column changes it calls `updateDatetimeValue` with the new value. The default handler for the cancel button sets it back to what it was:

https://github.com/ionic-team/ionic/blob/6e0b9c45489889266620ee2ca38c33fdf8ce3f3b/core/src/components/datetime/datetime.tsx#L305-L334

So not having this handler in the custom cancel button is the problem:

```javascript
{ 
  text: this.cancelText, 
  role: 'cancel', 
  handler: () => { 
    this.updateDatetimeValue(this.value); 
    this.ionCancel.emit(); 
  } 
}, 
```

cc @liamdebeasi

**Comment by Et3rnal** [comment]
@brandyscarney Can you provide an example? I tried setting the handler but it didn't work and still, the value gets changed by clicking cancel as well as clicking on the backdrop

Also, why we don't get the default buttons like in `"ionic-angular": "3.5.2"` adding the select and cancel buttons is so problematic !!!

I'm currently on `"@ionic/angular": "^4.7.1"`

**Comment by simonhaenisch** [comment]
@brandyscarney I haven't debugged this to dig deeper into the cause of the issue, but a pretty straightforward fix seems to be

```ts
if (!buttons || buttons.length === 0) {
  // add default buttons
} else {
  const cancelButton = buttons.find(button => button.role === 'cancel');

  if (cancelButton && !cancelButton.handler) {
    cancelButton.handler = () => { 
      this.updateDatetimeValue(this.value); 
      this.ionCancel.emit(); 
    }
}
```

However I'm not sure if this still behaves correctly when a custom cancel handler is passed, i. e., am I supposed to manually take care of not updating the value, and is it even possible to do so?

**Comment by brandyscarney** [comment]
I think this might require some rework in how the picker + datetime communicate, but I'm going to tag in @liamdebeasi since he's spent more time with datetime and see if he has ideas.

**Comment by Et3rnal** [comment]
@simonhaenisch your fix works just fine. However, I would suggest we add
```
}else if (cancelButton){
    this.ionCancel.emit();
}
```
As I think we should always trigger the event unless we can do that from outside?
This would help with what you mentioned about taking care of not updating as currently the cancel event doesn't get triggered and I'm struggling to find a workaround this issue because of that.

when can we expect a fit to this? would you like a PR?

**Comment by danbn-100295** [comment]
same issue,anything update ?
I can't add custom button with role cancel

**Comment by YoniZiv** [comment]
Same issue.

In my case, my  ion-datetime is binded to a ReactiveForm using the formControlName attribute.
I'm passing an object to pickerOptions property (on ion-datetime) which contains a buttons array with 2 objects.

If I pass them this way
![image](https://user-images.githubusercontent.com/20441936/80469037-aeb57680-8948-11ea-8f91-15017c77aa42.png)

then by selecting a date nothing happens. (even if I provide a role: 'cancel' property)

Once I add a handler to 'confirm me' it looks like this:

![image](https://user-images.githubusercontent.com/20441936/80469183-e3293280-8948-11ea-8019-b5c135b41b1f.png)

and from this point forward, this is what happens:

until I click 'confirm me' - clicking 'cancel me' would cancel.
once I click 'confirm me' - the value is set to my selection
from this point forward - clicking 'confirm me' or 'cancel me' would set the selected value.

remarks:
1) 'confirm handler called' is being logged only when clicking 'confirm me'.
2) if I provide the 'role=cancel' property - it's the same behavior
3) giving 'cancel me' a different/null handler doesn't fix the bug.

**Comment by TanmaySM** [comment]
Any Updates on this? Its still persist on Ionic 5.0.4.

**Comment by dingzhanjun** [comment]
this is still true with ionic/react 5.5.0

**Comment by ipacbalaj** [comment]
Do you have any update on this? 

As a quick fix I added an @ngIf on the component and I am reinitializing the compnent in the cancel handler.

![image](https://user-images.githubusercontent.com/8572545/110001639-3ca67e80-7d1d-11eb-8965-a46f655dcdf4.png)
![image](https://user-images.githubusercontent.com/8572545/110001674-46c87d00-7d1d-11eb-9787-6d421ad829c8.png)

**Comment by ihorbond** [comment]
Ionic team is notorious for not fixing bugs, why I'm not surprised smh seen same issues lingering all the way from v3 that still weren't fixed

**Comment by dharmeshds** [comment]
It's been a few years since reported first and I'm on the latest version of ionic/angular and still facing this bug! Is there ever gonna be a fix?

**Comment by liamdebeasi** [comment]
Hi everyone,

This is fixed in our new `ion-datetime` component coming in Framework v6. We are starting a beta process soon, so we will close issues like these out as we get closer to a stable release.

**Comment by dharmeshds** [comment]
Is there any workaround in Ionic5?

**Comment by liamdebeasi** [comment]
No, not at the moment.

**Comment by liamdebeasi** [comment]
Thanks for the issue. We released Ionic 6 in December of 2021, but it looks like we missed this issue when closing out resolved tickets. In Ionic 6 you now have more control over what the confirm/cancel buttons do in `ion-datetime`. See https://ionicframework.com/docs/api/datetime#buttons for more information.

I am going to close this as this reported issue has been resolved with this new architecture.
